zookeeper-server/src/main/java/org/apache/zookeeper/backup/NFSBackupStorage.java

```diff
@@ -0,0 +1,324 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Implementation for NFS-based backup storage provider
+ */
+public class NFSBackupStorage implements BackupStorageProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(NFSBackupStorage.class);
+
+  public static final String TMP_FILE_PREFIX = "TMP_";
+  private static final String PATH_SEPARATOR = File.separator;
+  private final BackupConfig _backupConfig;
+  private final String _fileRootPath;
+
+  /**
+   * Constructor using BackupConfig to get backup storage info
+   * @param backupConfig The information and settings about backup storage, to be set as a part of ZooKeeper server config
+   */
+  public NFSBackupStorage(BackupConfig backupConfig) {
+    _backupConfig = backupConfig;
+    _fileRootPath =
+        String.join(PATH_SEPARATOR, _backupConfig.getMountPath(), _backupConfig.getNamespace());
+  }
+
+  @Override
+  public BackupFileInfo getBackupFileInfo(File file) throws IOException {
+    String backupFilePath = constructBackupFilePath(file.getName());
+    File backupFile = new File(backupFilePath);
+
+    if (!backupFile.exists()) {
+      throw new BackupException(
+          "Backup file " + file.getName() + " does not exist, could not get file info.");
+    }
+
+    BasicFileAttributes fileAttributes =
+        Files.readAttributes(Paths.get(backupFilePath), BasicFileAttributes.class);
+    return new BackupFileInfo(backupFile, fileAttributes.lastModifiedTime().toMillis(),
+        fileAttributes.size());
+  }
+
+  @Override
+  public List<BackupFileInfo> getBackupFileInfos(File path, String prefix) throws IOException {
+    String backupDirPath = constructBackupFilePath(path.getPath());
+    File backupDir = new File(backupDirPath);
+
+    if (!backupDir.exists()) {
+      throw new BackupException(
+          "Backup directory " + path.getPath() + " does not exist, could not get file info.");
+    }
+
+    // Get a list of files whose file names start with the provided prefix
+    FilenameFilter fileFilter = (dir, name) -> name.startsWith(prefix);
+    File[] files = backupDir.listFiles(fileFilter);
+    if (files == null) {
+      throw new BackupException("The provided directory path " + path.getPath()
+          + " is invalid, could not get file info.");
+    }
```

Comment:
Why is it necessary that we throw an exception here instead of return an empty list?

Reply:
I thought returning an exception in that case might be an overkill - might require the caller to catch more exceptions. But if the path is not valid, it might be worth throwing an exception.

When would this path ever be not valid?

Reply:
The path is invalid means it is not we could not locate a file / directory using that path.